parsers/bibtex.c

```diff
@@ -0,0 +1,459 @@
+	Assert (isIdentChar (c));
+	do
+	{
+		vStringPut (string, c);
+		c = getcFromInputFile ();
+	} while (c != EOF && isIdentChar (c));
+	if (c != EOF)
+		ungetcToInputFile (c);		/* unget non-identifier character */
+}
+
+static bool readTokenFull (tokenInfo *const token, const bool includeWhitespaces)
+{
+	int c;
+	int whitespaces = -1;
+
+	token->type			= TOKEN_UNDEFINED;
+	token->keyword		= KEYWORD_NONE;
+	vStringClear (token->string);
+
+getNextChar:
+
+	do
+	{
+		c = getcFromInputFile ();
+		whitespaces++;
+	}
+	while (c == '\t'  ||  c == ' ' ||  c == '\n');
+
+	token->lineNumber   = getInputLineNumber ();
+	token->filePosition = getInputFilePosition ();
+
+	if (includeWhitespaces && whitespaces > 0 && c != '%' && c != EOF)
+	{
+		ungetcToInputFile (c);
+		c = ' ';
+	}
+
+	token->type = (unsigned char) c;
+	switch (c)
+	{
+		case EOF: return false;
+
+		case '@':
+				  /*
+				   * All Bib entries start with an at symbol.
+				   * Check if the next character is an alpha character
+				   * else it is not a potential tex tag.
+				   */
+				  c = getcFromInputFile ();
+				  if (! isalpha (c))
+					  ungetcToInputFile (c);
+				  else
+				  {
+					  vStringPut (token->string, '@');
+					  parseIdentifier (token->string, c);
+					  token->keyword = lookupKeyword (vStringValue (token->string) + 1, Lang_bib);
+					  if (isKeyword (token, KEYWORD_NONE)){
+						  token->type = TOKEN_IDENTIFIER;
+					  }else{
+						  token->type = TOKEN_KEYWORD;
+            }
+				  }
+				  break;
+		case '%':
+				  skipToCharacterInInputFile ('\n'); /* % are single line comments */
+				  goto getNextChar;
+				  break;
+
+		default:
+				  if (isIdentChar (c))
+				  {
+					  parseIdentifier (token->string, c);
+            token->type = TOKEN_IDENTIFIER;
+				  }
+				  break;
+	}
+	return true;
+}
+
+static bool readToken (tokenInfo *const token)
+{
+	return readTokenFull (token, false);
+}
+
+static void copyToken (tokenInfo *const dest, tokenInfo *const src)
+{
+	dest->lineNumber = src->lineNumber;
+	dest->filePosition = src->filePosition;
+	dest->type = src->type;
+	dest->keyword = src->keyword;
+	vStringCopy (dest->string, src->string);
+	vStringCopy (dest->scope, src->scope);
+}
+
+/*
+ *	 Scanning functions
+ */
+
+static bool parseTag (tokenInfo *const token, bibKind kind)
+{
+	tokenInfo *const name = newToken ();
+	vString *	currentid;
+	bool		useLongName = true;
+	bool        eof = false;
+
+	currentid = vStringNew ();
+	/*
+	 * Bib entries are of these formats:
+	 *   @article{identifier,
+   *   author="John Doe"}
+	 *
+	 * When a keyword is found, loop through all words up to
+	 * a comma brace for the tag name.
+	 *
+	 */
+	if (isType (token, TOKEN_KEYWORD))
+	{
+		copyToken (name, token);
+		if (!readToken (token))
+		{
+			eof = true;
+			goto out;
+		}
+	}
+
+	if (isType (token, TOKEN_OPEN_CURLY))
+	{
+		int depth = 1;
+
+		if (!readToken (token))
+		{
+			eof = true;
+			goto out;
+		}
+    if (isType (token, TOKEN_IDENTIFIER)){
```

Comment:
It seems that indentation is a bit broken.

Reply:
also 952f651